File: LeetCode/Kai/230KthSmallestElementInaBST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int kthSmallest(TreeNode root, int k) {
        Stack<TreeNode> stack = new Stack<TreeNode>();
        int count = 0;
        TreeNode curr = root;
        while(true) {
            while(curr!=null) {
                stack.push(curr);
                curr = curr.left;
            }
            curr = stack.pop();
            count++;
            if(count == k) return curr.val;
            curr = curr.right;
        }
    }
    
}

File: LeetCode/Liang/230KthSmallestElementInABST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 
/**
 * My time complexity is o(n), where n is the total number of the node.
 **/
 
// static int countNum = 0; This will definitely have problem when the class is used in multipal test.
// static int ans = 0;
// int countNum = 0;
// int ans = 0;
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        kthSmallestHelper(root, k);
        return ans;
    }
private:
    int ans;
    int countNum = 0;
    void kthSmallestHelper(TreeNode *root, int k) {
        if (root->left) {
            kthSmallestHelper(root->left, k);
        }
        ++countNum;
        if (countNum == k) {
            ans = root->val;
            return; // Here is good.
        }
        if (root->right) {
            kthSmallestHelper(root->right, k);
        }
    }
};
